File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<pair<char,int>> charStack;
        string answer = "";  
        string tmp = ""; 
        for(int i = 0; i < s.length(); i++){
            if(isalpha(s[i])){
                tmp += s[i]; 
            } else if(s[i] == '('){
                charStack.push({s[i],i}); 
                tmp += '?'; 
            } else{
                if(!charStack.empty()){
                    tmp += ')';
                    tmp[charStack.top().second] = '(';  
                    charStack.pop(); 
                } else{
                    tmp += '!'; 
                }
            }
        }

        for(char& c : tmp){
            if(c != '?' && c != '!'){
                answer += c; 
            }
        }
        return answer; 
    }
};

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        Deque<Integer> stack = new ArrayDeque<>(); 
        Set<Integer> banned = new HashSet<>(); 
        for(int i = 0; i < s.length(); i++){
            char curr = s.charAt(i); 
            if(curr == ')'){
                if(!stack.isEmpty()) stack.pop(); 
                else banned.add(i); 
            } else if(curr == '('){
                stack.push(i); 
            }
        }

        while(!stack.isEmpty()){
            banned.add(stack.pop()); 
        }

        String answer = ""; 
        for(int i = 0; i < s.length(); i++){
            if(!banned.contains(i)){
                answer += s.charAt(i); 
            }
        }

        return answer; 
    }
}